python/ray/tune/experimental/output.py

```diff
@@ -0,0 +1,585 @@
+        if current_best_trial:
+            result.append(_best_trial_str(current_best_trial, metric))
+        # Now populating the trial table data.
+        if not self._inferred_metric:
+            # try inferring again.
+            self._inferred_metric = _infer_user_metrics(trials)
+
+        all_metrics = list(DEFAULT_COLUMNS.keys()) + self._inferred_metric
+
+        trial_table_data = _get_progress_table_data(trials, all_metrics)
+        return result, trial_table_data
+
+    def _print_heartbeat(self, trials, *sys_args):
+        raise NotImplementedError
+
+
+class TuneTerminalReporter(TuneReporterBase):
+    def _print_heartbeat(self, trials, *sys_args):
+        if self._verbosity < self._heartbeat_threshold:
+            return
+        heartbeat_strs, table_data = self._get_heartbeat(trials, sys_args)
+        for s in heartbeat_strs:
+            print(s)
+        # now print the table using Tabulate
+        header = table_data.header
+        table_data_list = table_data.data
+        header_printed = False
+        for table in table_data_list:
+            if not header_printed:
+                print(
+                    tabulate(
+                        table.trial_infos,
+                        headers=header,
+                        tablefmt="simple",
+                        showindex=False,
+                    )
+                )
+                header_printed = True
+                if table.more_info:
+                    print(table.more_info)
+            else:
+                print(tabulate(table.trial_infos, tablefmt="simple", showindex=False))
+
+
+class TuneRichReporter(TuneReporterBase):
+    def _print_heartbeat(self, trials, *args):
+        # TODO: display `self._get_heartbeat` using Rich
+        pass
+
+
+class TrainReporter(ProgressReporter):
+    # the minimal verbosity threshold at which heartbeat starts getting printed.
+    _heartbeat_threshold = AirVerbosity.VERBOSE
+
+    def _get_heartbeat(self, trials: List[Trial]):
+        # Iteration 2 is running at 2023-02-24 12:35:39. Running time: 42min 14s
+        if len(trials) == 0:
+            return
+        trial = trials[0]
+        if trial.status != Trial.RUNNING:
+            return " ".join(
+                [f"Training is in {trial.status} status.", self._time_passed_str]
+            )
+        if not trial.last_result:
+            iter_num = 1
+        else:
+            iter_num = trial.last_result[TRAINING_ITERATION] - 1
+        return " ".join([f"Training on iteration {iter_num}.", self._time_passed_str])
+
+    def _print_heartbeat(self, trials, *args):
+        print(self._get_heartbeat(trials))
+
+
+# These keys are blacklisted for printing out training/tuning intermediate/final result!
+BLACKLISTED_KEYS = {
+    "date",
+    "done",
+    "hostname",
+    "iterations_since_restore",
+    "node_ip",
+    "pid",
+    "time_since_restore",
+    "timestamp",
+    "trial_id",
+    "experiment_tag",
+    "should_checkpoint",
```

Comment:
why?

Reply:
all these clutter our output - they are not that useful to our users, some of them are autofilled by us.

Reply:
for example, trial_id is already obvious from
```
TRIAL_NAME (5 iters) finished at at 2023-02-24 12:35:39. Running time: 2min 14s
```